pyls/plugins/completion.py

```diff
@@ -3,28 +3,89 @@
 from pyls.lsp import CompletionItemKind
 from pyls import hookimpl
 
+from threading import Thread, Lock
+from rope.contrib.codeassist import code_assist, sorted_proposals
+
 log = logging.getLogger(__name__)
 
 
+class CompletionThread(Thread):
```

Comment:
It might also be nice to generalise this somehow. Ideally, we could call any of the pluggy hooks in a "race" mode where the first non-None response wins.

Reply:
@gatesn Would you like to extract these kind of methods to anywhere else?

Reply:
Yes please. I think the best way to structure this is to rename this to `jedi_completion.py` and create a new `rope_completion.py`. Then in `python_ls.py` you can grab the hooks from pluggy and run them through this "racer" that runs all hooks and returns the first result.

Reasoning is:

* There may be more things in the future that we want to race like this.
* This helps us step towards the LSP streaming extension (return one set of results, then update them some time later)
* People can explicitly choose to disable/enable rope or jedi if they have any problems